Handle exceptions in subscribe callbacks

arkivanov opened the pull request:
Quite a big change, extracted boilerplate to a separate file.
Fixes #176

CherryPerry commented on reaktive/src/commonMain/kotlin/com/badoo/reaktive/utils/Utils.kt:
```diff
@@ -10,7 +12,8 @@ internal fun handleSourceError(error: Throwable, onError: ((Throwable) -> Unit)?
             } catch (e: Throwable) {
```
Please, rename `e` into `errorHandlerException` or whatever. It is hard to distinct them.

CherryPerry commented on reaktive/src/commonMain/kotlin/com/badoo/reaktive/utils/PrintError.kt:
```diff
@@ -1,3 +1,14 @@
+@Suppress("ObjectPropertyName") // Backing field
+private val _isPrintErrorEnabled = AtomicBoolean(true)
+
+internal var isPrintErrorEnabled
```
Use it for `printStack()` too. It is quite strange, that you can skip some logging of small messages like `Some shit happened` but not huge stacktraces.

Maybe in future we should introduce `Logger` that can log text and exceptions (with stacktraces) and default implementation of it in the way it is implemented now.
For example on Android it will become possible to use `Tinder` to log it and report to server.

CherryPerry replied:
Let's keep it as is for now, but introduce `Logger ` interface later, so we can mock logger for android.

arkivanov commented:
@CherryPerry Added more tests, found and fixed a few bugs.

arkivanov commented:
iOS build failed as always. Will fix using logs..............

arkivanov commented:
Added tests for `currentThreadId` to make sure it returns same values for same threads and different values for different threads.

arkivanov commented:
Also moved `doInBackground` from `commonTest` to `jvmNativeCommonTest` and split for Native and JVM. Because scheduler based approach reuses threads but for tests we need new threads every time.
